File: Graph/Medium/Lnt-Q127-Topological-Sorting/solution_bfs.cpp
class Solution {
public:
    /*
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */
    vector<DirectedGraphNode*> topSort(vector<DirectedGraphNode*>& graph) {
        // bfs + hash table
	   //TC:O(m*n); SC:O(n)
        if(graph.size() == 0) {
            return {};
        }
        unordered_map<DirectedGraphNode*, int> dict;
        vector<DirectedGraphNode*> res;
        queue<DirectedGraphNode*> q;
        for(auto node : graph) { // note all the nodes with in-dgree >= 1
            for(auto nd : node->neighbors) {
                dict[nd] = dict.find(nd) != dict.end() ? ++dict[nd] : 1;
            }
        }
        for(auto node: graph) {
            if(dict.find(node) == dict.end()) { // find all nodes with in-dgree = 0
                q.push(node); // add to queue
            }
        }
        while(!q.empty()) {
            auto node = q.front();
            q.pop();
            res.push_back(node);
            for(auto nd : node->neighbors) {
                dict[nd]--;  // has been accessed, in-degree--
                if(dict[nd] == 0) { // in-degree = 0 => add to the queue
                    q.push(nd);
                }
            }
        }
        return res;
    }
};

File: Graph/Medium/210-Course-Schedule-II/solution_bfs.cs
public class Solution {
    public int[] FindOrder(int numCourses, int[,] prerequisites) {
        // bfs + hash table
        // tc:O(mn); sc:O(n)
        if(numCourses == 0) {
            return new int[0];
        }
        Dictionary<int, int> dict = new Dictionary<int, int>(); // record all the in-degree of node
        Dictionary<int, IList<int>> graph = new Dictionary<int, IList<int>>(); // build the node with edges
        List<int> res = new List<int>();
        Queue<int> queue = new Queue<int>();
        int sum = 0;
        for(int i = 0; i < prerequisites.GetLength(0); i++) {
            dict[prerequisites[i,0]] = dict.ContainsKey(prerequisites[i,0]) ? ++dict[prerequisites[i,0]] : 1; // calculate in-dgree
            if(graph.ContainsKey(prerequisites[i,1])) { //build the graph
                graph[prerequisites[i,1]].Add(prerequisites[i,0]);
            }
            else {
                graph[prerequisites[i,1]] = new List<int>{prerequisites[i,0]}; 
            }
        }     
        for(int i = 0; i < numCourses; i++) { // find all node with in-degree = 0
            if(!dict.ContainsKey(i)) {
                queue.Enqueue(i); // push to queue
            }
        }
        while(queue.Count > 0) {
            int node = queue.Dequeue();
            sum++;
            res.Add(node);
            if(graph.ContainsKey(node)){
                foreach(var course in graph[node]) {
                    dict[course]--;
                    if(dict[course] == 0) {
                        queue.Enqueue(course);
                    }
                }
            }
        }
        return sum == numCourses ? res.ToArray() : new int[0];
    }
}